crates/sui-move-build/src/linters/freeze_wrapped.rs

```diff
@@ -0,0 +1,259 @@
+) {
+    use H::Statement_ as S;
+    match stmt {
+        S::Command(cmd) => visit_cmd(cmd, env, modules),
+        S::IfElse {
+            cond,
+            if_block,
+            else_block,
+        } => {
+            visit_exp(cond, env, modules);
+            visit_block(if_block, env, modules);
+            visit_block(else_block, env, modules);
+        }
+        S::While {
+            cond: (c, b),
+            block,
+        } => {
+            visit_block(c, env, modules);
+            visit_exp(b, env, modules);
+            visit_block(block, env, modules);
+        }
+        S::Loop {
+            block,
+            has_break: _,
+        } => visit_block(block, env, modules),
+    }
+}
+
+fn visit_block(
+    block: &H::Block,
+    env: &mut CompilationEnv,
+    modules: &UniqueMap<E::ModuleIdent, H::ModuleDefinition>,
+) {
+    for s in block {
+        visit_stmt(s, env, modules);
+    }
+}
+
+fn visit_cmd(
+    sp!(_, cmd): &H::Command,
+    env: &mut CompilationEnv,
+    modules: &UniqueMap<E::ModuleIdent, H::ModuleDefinition>,
+) {
+    use H::Command_ as C;
+    match cmd {
+        C::Assign(_, e) => visit_exp(e, env, modules),
+        C::Mutate(e1, e2) => {
+            visit_exp(e1, env, modules);
+            visit_exp(e2, env, modules);
+        }
+        C::Abort(e) => visit_exp(e, env, modules),
+        C::Return { from_user: _, exp } => visit_exp(exp, env, modules),
+        C::IgnoreAndPop { pop_num: _, exp } => visit_exp(exp, env, modules),
+        C::JumpIf {
+            cond,
+            if_true: _,
+            if_false: _,
+        } => visit_exp(cond, env, modules),
+        C::Break | C::Continue | C::Jump { .. } => (),
+    }
+}
+
+fn visit_exp(
+    exp: &H::Exp,
+    env: &mut CompilationEnv,
+    modules: &UniqueMap<E::ModuleIdent, H::ModuleDefinition>,
+) {
+    use H::UnannotatedExp_ as E;
+    let sp!(_, uexp) = &exp.exp;
+    match uexp {
+        E::ModuleCall(fun) => {
+            if FREEZE_FUNCTIONS
+                .iter()
+                .any(|(addr, module, fname)| fun.is(addr, module, fname))
+            {
+                // single argument passed by value
+                let H::Type_::Single(st) = &fun.arguments.ty.value else {
+                    return;
+                };
+                let bt = match &st.value {
+                    H::SingleType_::Base(bt) => bt,
+                    H::SingleType_::Ref(_, bt) => bt,
+                };
+                let H::BaseType_::Apply(_, tname ,_ ) = &bt.value else {
+                    return;
+                };
+                if let H::TypeName_::ModuleType(mident, sname) = tname.value {
+                    if let Some((sfields, sloc)) = struct_fields(&sname, &mident, modules) {
+                        for (f, t) in sfields {
+                            if let Some((nested_sname, nested)) =
+                                contains_key(t, modules, /*field_depth*/ 0)
```

Comment:
This feels a bit wasteful to check everytime? I feel like you should be able to pre-populate all of this information for every time before visiting expressions. In other words:
- Visit all structs across all modules
  - You will need to do some sort of saturating pass for this I think? Which is maybe why you took this approach you have? 
- Record which have fields with key 
- Visit exps for freeze

Reply:
I would advocate for the approach I laid out over some sort of memoized layer on top of what has been done here in that I would hope what I have suggested is more parallelizable in the future 
(it might not look it, but other than a few annoying spots with Context, the pass structure of the compiler is designed to be highly amenable to parallelization (which we need to do some day))

Reply:
Done!